### PR TITLE
Replace parseFullName with parseGoObject

### PR DESCRIPTION
**Pull request by Aflynn50**
This PR replaces parseFullName with the more specific parseGoObject. Previously, parseFullName was intended for use parsing columns and go objects but we decided it made more sense to have two separate functions.

By using parseGoObject we can also be more specific with how we parse input expressions. We will parse an input expression as long as the character after the $ is a name byte. We ignore the character before the $.

The errors returned by parseInputExpression also now include an approximate position of the error. We intend to provide an approximate position for all errors in future PRs.

### CODE REVIEW

File: internal/parse/parser.go [hunk 1]
```diff
@@ -250,51 +243,67 @@ func (p *Parser) parseIdentifier() (string, bool) {
...
-				return fn, true
-			}
-		} else {
-			// A column name specified without a table prefix is a name not a
-			// prefix.
-			if idc == columnId {
-				fn.Name = fn.Prefix
-				fn.Prefix = ""
+			if idField, ok := p.parseIdentifier(allowAsterisk); ok {
+				return FullName{Prefix: id, Name: idField}, true, nil
```
Comment (fernape) [inline internal/parse/parser.go]: We can use the short form here: `return FullName{id, idField}, true, nil`

File: internal/parse/parser.go [hunk 2]
```diff
@@ -250,51 +243,67 @@ func (p *Parser) parseIdentifier() (string, bool) {
...
+		if fn.Name == "*" {
+			s++
+		}
+	}
+	return s
 }
 
 // parseInputExpression parses an input expression of the form $Type.name.
-func (p *Parser) parseInputExpression() (*InputPart, bool, error) {
+func (p *Parser) parseInputExpression() (ip *InputPart, ok bool, err error) {
```
Comment (fernape) [inline internal/parse/parser.go]: I am not sure about the named return values. `ip` and `ok` are never "used" since the function always composes the `return` statement with hardcoded first and second parameters. Using `:=` in `if fn, ok, err = p.parseGoObject(); ok {` would have the same result. In addition we'd need to remove `var fn FullName`. I don't know out policy about this actually.

Comment (Aflynn50) [reply internal/parse/parser.go]: Removed.

File: internal/parse/parser.go (hunk 1 above)
Comment (Aflynn50) [reply internal/parse/parser.go]: Changed.

File: internal/parse/parser.go [hunk 3]
```diff
@@ -243,58 +236,64 @@ func (p *Parser) skipName() bool {
...
-func (p *Parser) parseIdentifier() (string, bool) {
-	if p.skipByte('*') {
-		return "*", true
-	}
+type asteriskFlag int
+
+const (
+	allowAsterisk asteriskFlag = iota
+	disallowAsterisk
+)
```
Comment (TheSignPainter98) [inline internal/parse/parser.go]: Instead of `asteriskFlag` being a separate type, could you use a `bool`?

File: internal/parse/parser.go [hunk 4]
```diff
@@ -243,58 +236,64 @@ func (p *Parser) skipName() bool {
...
 
-	idStart := p.pos
+// parseIdentifier parses either a name made up only of nameBytes (or an
+// asterisk if asteriskF=allowAsterisk). On success it returns the parsed string
+// and true. Otherwise, it returns the empty string and false.
```
Comment (TheSignPainter98) [inline internal/parse/parser.go]: The mention of `asteriskF` is surprising, also the 'either' dangles if you ignore the brackets. Perhaps something like:
```
parseIdentifier parses a name made up of only nameBytes, or alternatively
a single asterisk if `allowAsterisk` is passed.
```
would be cleaner?

File: internal/parse/parser.go [hunk 5]
```diff
@@ -243,58 +236,64 @@ func (p *Parser) skipName() bool {
...
+func (p *Parser) parseIdentifier(asteriskF asteriskFlag) (string, bool) {
+	if asteriskF == allowAsterisk {
+		if p.skipByte('*') {
+			return "*", true
+		}
+	}
+	mark := p.pos
 	if p.skipName() {
-		return p.input[idStart:p.pos], true
+		return p.input[mark:p.pos], true
```
Comment (TheSignPainter98) [inline internal/parse/parser.go:255]: `idStart` seems clearer to me

File: internal/parse/parser.go [hunk 6]
```diff
@@ -243,58 +236,64 @@ func (p *Parser) skipName() bool {
...
 }
 
-// parseFullName parses a column name or Go field name, optionally dot-prefixed by
-// its table or type name respectively.
-func (p *Parser) parseFullName(idc idClass) (FullName, bool) {
+// parseGoObject parses a source or target go object of the form Prefix.Name
+// where the type is the Prefix and the field is the Name (this applies to maps
+// and structs). On sccuess it returns the parsed FullName, true and nil.
+// If it cannot parse the Go object it returns false. If the construct was
+// recognised as a Go object there will be an error. Otherwise err=nil.
```
Comment (TheSignPainter98) [inline internal/parse/parser.go]: `s/sccuess/success/`

Also 'if the construct _was_ recognised...' should this be _was not?_ :)
More generally though, I think the last two lines can be simplified to:
```
If it cannot parse the Go object it returns false and an error.
```

Comment (Aflynn50) [reply internal/parse/parser.go]: I think this comment wasn't very clear, I've reworded it.
```If a Go object is found, but not formatted correctly, false and an error are returned. Otherwise the error is nil.```

File: internal/parse/parser.go (hunk 5 above)
Comment (Aflynn50) [reply internal/parse/parser.go:255]: It was changed to be consistent with the other skip functions. Hopefully it's still easy to understand.

File: internal/parse/parser.go (hunk 4 above)
Comment (Aflynn50) [reply internal/parse/parser.go]: Changed.

File: internal/parse/parser.go (hunk 3 above)
Comment (Aflynn50) [reply internal/parse/parser.go]: Using a bool was considered, but we opted to go for a const to make it clearer in functions where `parseIdentifier` is used. This way it is clear to see that the part before the `.` in a Go object cannot be an asterisk but the part after can. (It would be so nice to have a Haskell style two element type with named members but alas)

Comment (TheSignPainter98) [reply internal/parse/parser.go]: Ah okay I understand, thanks

File: internal/parse/parser.go (hunk 5 above)
Comment (TheSignPainter98) [reply internal/parse/parser.go:255]: I wasn't aware of this context, conforming to the pattern seems like a good shout though

File: internal/parse/parser.go [hunk 7]
```diff
@@ -243,58 +236,64 @@ func (p *Parser) skipName() bool {
...
-				fn.Prefix = ""
-			}
-			return fn, true
+			return fn, false, fmt.Errorf("invalid identifier near char %d", p.pos)
 		}
+		return fn, false, fmt.Errorf("go object near char %d not qualified", p.pos)
 	}
 	cp.restore()
-	return fn, false
+	return fn, false, nil
```
Comment (solmyrcan) [inline internal/parse/parser.go]: Could this method return `nil, false, nil` in failure cases instead of `fn, false, nil` because with this change, `fn` seems to be no longer used. We could delete the declaration for `fn` at the top as well :)

Comment (Aflynn50) [reply internal/parse/parser.go]: Thats a good point. Go doesnt actually allow us to use nil there because its not a zero value for structs, but I have just changed it to return `FullName{}` instead so I can remove the var declaration.

File: internal/parse/parser.go [hunk 8]
```diff
@@ -243,58 +236,65 @@ func (p *Parser) skipName() bool {
...
+const (
+	allowAsterisk asteriskFlag = iota
+	disallowAsterisk
+)
+
+// parseIdentifier parses a name made up of only nameBytes, or alternatively a
+// single asterisk if `allowAsterisk` is passed. On success it returns the
+// parsed string and true. Otherwise, it returns the empty string and false.
+func (p *Parser) parseIdentifier(asteriskF asteriskFlag) (string, bool) {
+	if asteriskF == allowAsterisk {
```
Comment (niemeyer) [inline internal/parse/parser.go]: Instead of having a flag, why not just having two functions:
```
func parseAsteriskIdentifier() (string, bool) {
     if p.skipByte(...) {
         return "*", ...
     }
    return parseIdentifier()
}

func parseIdentifier() (string, bool) {
    return
}
```

Up to you, though.

File: internal/parse/parser.go [hunk 9]
```diff
@@ -243,58 +236,65 @@ func (p *Parser) skipName() bool {
...
 
-// parseFullName parses a column name or Go field name, optionally dot-prefixed by
-// its table or type name respectively.
-func (p *Parser) parseFullName(idc idClass) (FullName, bool) {
+// parseGoObject parses a source or target go object of the form Prefix.Name
+// where the type is the Prefix and the field is the Name (this applies to maps
+// and structs). On success it returns the parsed FullName, true and nil. If a
+// Go object is found, but not formatted correctly, false and an error are
+// returned. Otherwise the error is nil.
+func (p *Parser) parseGoObject() (FullName, bool, error) {
```
Comment (niemeyer) [inline internal/parse/parser.go]: This looks fine, but wasn't the point of having a parseFullName that "database.column" and "Type.db_field" look the same?

File: internal/parse/parser.go [hunk 10]
```diff
@@ -243,58 +236,65 @@ func (p *Parser) skipName() bool {
...
-		return p.input[idStart:p.pos], true
+		return p.input[mark:p.pos], true
 	}
 	return "", false
 }
 
-// parseFullName parses a column name or Go field name, optionally dot-prefixed by
-// its table or type name respectively.
-func (p *Parser) parseFullName(idc idClass) (FullName, bool) {
+// parseGoObject parses a source or target go object of the form Prefix.Name
```
Comment (niemeyer) [inline internal/parse/parser.go]: It says Prefix.Name, but assuming this remains as parseGoObject, isn't this "TypeName.db_tag_name"?  In other words, the db_tag_name will typically be lower case and not TitleCased because it's supposed to match a column name in the database. If that's true, then the name parseGoObject might also require some thinking as this isn't a "Go object", strictly speaking.

**Review by niemeyer — CHANGES_REQUESTED** [review]
Logic looks reasonable. Question is mainly about details and the background behind these functions.

File: internal/parse/parser.go (hunk 8 above)
Comment (Aflynn50) [reply internal/parse/parser.go]: Changed. We were trying to work out the most idiomatic way to do this but I think separate functions is nicer than what we had.

Comment (Aflynn50) [reply internal/parse/parser.go]: Changed. We were trying to work out the most idiomatic way to do this but I think separate functions is nicer than what we had.

File: internal/parse/parser.go (hunk 9 above)
Comment (Aflynn50) [reply internal/parse/parser.go]: If we want to support function columns and columns without table prefixes it makes more sense to have separate functions. It also allows us to give better errors. It would be possible to parse them in one function but it would be hard to follow and rely on a lot of conditionals, even if it was more compact.

File: internal/parse/parser.go (hunk 10 above)
Comment (Aflynn50) [reply internal/parse/parser.go]: Changed.

File: internal/parse/parser.go [hunk 11]
```diff
@@ -243,58 +236,62 @@ func (p *Parser) skipName() bool {
...
 }
 
-// parseFullName parses a column name or Go field name, optionally dot-prefixed by
-// its table or type name respectively.
-func (p *Parser) parseFullName(idc idClass) (FullName, bool) {
+// parseGoTarget parses a go type name qualified by a tag name (or asterisk) of
+// the form "TypeName.col_name". On success it returns the parsed FullName, true
+// and nil. If a target is found, but not formatted correctly, false and an
+// error are returned. Otherwise the error is nil.
+func (p *Parser) parseGoTarget() (FullName, bool, error) {
```
Comment (niemeyer) [inline internal/parse/parser.go]: Isn't this both Source and Target?  If so, parseGoFullName might be more appropriate. If not, then it's fine.

File: internal/parse/parser_test.go [hunk 12]
```diff
@@ -191,28 +195,34 @@ func TestRound(t *testing.T) {
...
-	parser := parse.NewParser()
-	_, err := parser.Parse(sql)
-	assert.Equal(t, fmt.Errorf("cannot parse expression: missing right quote in string literal"), err)
-}
+	var testTable = []struct {
+		name string
+		sql  string
+		err  string
+	}{
+		{
```
Comment (niemeyer) [inline internal/parse/parser_test.go]: The two parenthesis above can be merged together. That's below as well on the ending (line 217).

This is the same style used right above in this same diff.

File: internal/parse/parser_test.go [hunk 13]
```diff
@@ -231,58 +241,36 @@ func TestBadEscaped(t *testing.T) {
...
-	parser := parse.NewParser()
-	_, err := parser.Parse(sql)
-	assert.Equal(t, fmt.Errorf("cannot parse expression: malformed input type"), err)
-}
+	var testTable = []struct {
+		name string
+		sql  string
+		err  string
+	}{
+		{
```
Comment (niemeyer) [inline internal/parse/parser_test.go]: Same here.

**Review by niemeyer — APPROVED** [review]
Only nitpicks now. LGTM, thanks.

File: internal/parse/parser.go (hunk 11 above)
Comment (Aflynn50) [reply internal/parse/parser.go]: Changed.

File: internal/parse/parser_test.go (hunk 12 above)
Comment (Aflynn50) [reply internal/parse/parser_test.go]: Changed.

File: internal/parse/parser_test.go (hunk 13 above)
Comment (Aflynn50) [reply internal/parse/parser_test.go]: Changed.
